File: Algorithms/Dynamic_Programming/Target_Sum/Target_Sum.c
#include <stdio.h>
#include <stdlib.h> 

int CntSum(int arr[], int N, int required_sum) 
{ 
    int t[N+1][required_sum+1] , i ,j ;
    for(i = 0 ; i < N+1;i++)
        t[i][0] = 1;
    for(i = 1 ; i < required_sum+1;i++)
        t[0][i] = 0;
    for(i = 1; i < N+1;i++)
    {
        for(j = 1 ; j < required_sum+1;j++)
        {
            if(arr[i-1] <= j)
                t[i][j] = t[i-1][j-arr[i-1]]+t[i-1][j];
            else
                t[i][j] = t[i-1][j];
        }
    }
    return t[N][required_sum]; 
} 
int main() 
{ 
    int N , i;
    printf("Enter Size of Array :- \n");
    scanf("%d",&N);
    int A[N];
    printf("Enter Array Elements :- \n");
    for(i = 0 ; i< N;i++)
        scanf("%d",&A[i]);
    printf("Enter Target Sum :- \n");
    int X;
    scanf("%d",&X);
    long int ans = CntSum(A , N , X);
    printf("Count = %ld\n", ans);
  
    return 0; 
} 

File: Algorithms/DP/Target_Sum/Target_Sum.py
arr = list(map(int, input('Enter array: ').split()))
targetSum = int(input('Enter target sum: '))

dp = [[0 for i in range(targetSum + 1)] for i in range(len(arr))]
visited = [[0 for i in range(targetSum + 1)] for i in range(len(arr))]

def findNumberOfSubsetsMemo(arr, i, targetSum, n) : 
    
    print(i, targetSum)

    # Base condition
    if (i == n) : 
        if (targetSum == 0) : 
            return 1;  
        else : 
            return 0;  

    if(targetSum < 0):

        return 0;
   
    #Retrieve if already solved
    if (visited[i][targetSum] == 1) : 
        return dp[i][targetSum];  
    
    #Set as solved
    visited[i][targetSum] = 1;  
  
    # Recursion call
    dp[i][targetSum] = findNumberOfSubsetsMemo(arr, i + 1, targetSum, n) + findNumberOfSubsetsMemo(arr, i + 1, targetSum - arr[i], n);  
      
    return dp[i][targetSum];

def findNumberOfSubsetsRecur(arr, n, i,sum, count): 
      
    if (i == n): 
  
        if (sum == 0): 
            count += 1
        return count 
  
    count = findNumberOfSubsetsRecur(arr, n, i + 1, sum - arr[i], count) 
    count = findNumberOfSubsetsRecur(arr, n, i + 1, sum, count) 
    return count

print('By Memo:', findNumberOfSubsetsMemo(arr, 0, targetSum, len(arr)))